File: packages/app/android/src/reactnative/java/io/invertase/firebase/common/ReactNativeFirebaseModule.java
package io.invertase.firebase.common;

/*
 * Copyright (c) 2016-present Invertase Limited & Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this library except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import android.app.Activity;
import android.content.Context;
import androidx.annotation.CallSuper;
import androidx.annotation.NonNull;
import com.facebook.react.bridge.*;
import io.invertase.firebase.interfaces.ContextProvider;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ExecutorService;

public class ReactNativeFirebaseModule extends ReactContextBaseJavaModule
    implements ContextProvider {
  private final TaskExecutorService executorService;

  private String moduleName;

  public ReactNativeFirebaseModule(ReactApplicationContext reactContext, String moduleName) {
    super(reactContext);
    this.moduleName = moduleName;
    this.executorService = new TaskExecutorService(getName());
  }

  public static void rejectPromiseWithExceptionMap(Promise promise, Exception exception) {
    promise.reject(exception, SharedUtils.getExceptionMap(exception));
  }

  public static void rejectPromiseWithCodeAndMessage(
      Promise promise, String code, String message, ReadableMap resolver) {
    WritableMap userInfoMap = Arguments.createMap();
    userInfoMap.putString("code", code);
    userInfoMap.putString("message", message);
    if (resolver != null) {
      userInfoMap.putMap("resolver", resolver);
    }
    promise.reject(code, message, userInfoMap);
  }

  public static void rejectPromiseWithCodeAndMessage(Promise promise, String code, String message) {
    WritableMap userInfoMap = Arguments.createMap();
    userInfoMap.putString("code", code);
    userInfoMap.putString("message", message);
    promise.reject(code, message, userInfoMap);
  }

  public static void rejectPromiseWithCodeAndMessage(
      Promise promise, String code, String message, String nativeErrorMessage) {
    WritableMap userInfoMap = Arguments.createMap();
    userInfoMap.putString("code", code);
    userInfoMap.putString("message", message);
    userInfoMap.putString("nativeErrorMessage", nativeErrorMessage);
    promise.reject(code, message, userInfoMap);
  }

  @Override
  @CallSuper
  public void initialize() {
    super.initialize();
  }

  public ReactContext getContext() {
    return getReactApplicationContext();
  }

  public final ExecutorService getExecutor() {
    return executorService.getExecutor();
  }

  public final ExecutorService getTransactionalExecutor() {
    return executorService.getTransactionalExecutor();
  }

  public final ExecutorService getTransactionalExecutor(String identifier) {
    return executorService.getTransactionalExecutor(identifier);
  }


  // This is no longer called as of react-native 0.74 and is only here for
  // compatibility with older versions. It delegates to thew new `invalidate`
  // method, which all modules should implement now
  // Remove this method when minimum supported react-native is 0.74
  /** @noinspection removal*/
  @SuppressWarnings({"deprecation", "removal"})
  @Deprecated
  public void onCatalystInstanceDestroy() {
    // This should call the child class invalidate, which will then call super.invalidate,
    // and everything will work correctly up and down the inheritance hierarchy to shut down
    invalidate();
  }

  // This should have an @Override annotation but we cannot do
  // that until our minimum supported react-native version is 0.74, since the
  // method did not exist before then
  @CallSuper
  public void invalidate() {
    super.invalidate();
    executorService.shutdown();
  }

  public final void removeEventListeningExecutor(String identifier) {
    String executorName = executorService.getExecutorName(true, identifier);
    executorService.removeExecutor(executorName);
  }

  public Context getApplicationContext() {
    return getReactApplicationContext().getApplicationContext();
  }

  public Activity getActivity() {
    return getCurrentActivity();
  }

  @NonNull
  @Override
  public String getName() {
    return "RNFB" + moduleName + "Module";
  }

  @NonNull
  @Override
  public Map<String, Object> getConstants() {
    return new HashMap<>();
  }
}

File: packages/app/android/src/reactnative/java/io/invertase/firebase/common/ReactNativeFirebaseModule.kt
package io.invertase.firebase.common

/*
 * Copyright (c) 2016-present Invertase Limited & Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this library except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import android.app.Activity
import android.content.Context
import androidx.annotation.CallSuper
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReadableMap
import io.invertase.firebase.interfaces.ContextProvider
import java.util.HashMap
import java.util.concurrent.ExecutorService

open class ReactNativeFirebaseModule(
  reactContext: ReactApplicationContext?,
  private var moduleName: String?,
) : ReactContextBaseJavaModule(reactContext),
  ContextProvider {
  private val executorService = TaskExecutorService(name)

  @CallSuper
  override fun initialize() {
    super.initialize()
  }

  override fun getContext(): ReactContext? = reactApplicationContext

  fun getExecutor(): ExecutorService = executorService.getExecutor()

  fun getTransactionalExecutor(): ExecutorService = executorService.getTransactionalExecutor()

  fun getTransactionalExecutor(identifier: String): ExecutorService = executorService.getTransactionalExecutor(identifier)

  /**
   * Kept for source and runtime compatibility with React Native versions that called this method.
   * Dynamic dispatch is intentional: a child invalidate implementation must run before it calls
   * back into [invalidate].
   */
  @Suppress("DEPRECATION", "OVERRIDE_DEPRECATION")
  @java.lang.Deprecated
  @Deprecated("Use invalidate() instead.")
  override fun onCatalystInstanceDestroy() {
    invalidate()
  }

  /**
   * Current React Native defines invalidate as an empty subclass hook. Do not call super here:
   * older supported React Native versions delegated back to onCatalystInstanceDestroy, which
   * would recurse through this compatibility method.
   */
  @CallSuper
  override fun invalidate() {
    executorService.shutdown()
  }

  fun removeEventListeningExecutor(identifier: String) {
    val executorName = executorService.getExecutorName(true, identifier)
    executorService.removeExecutor(executorName)
  }

  override fun getApplicationContext(): Context? = reactApplicationContext.applicationContext

  override fun getActivity(): Activity? = reactApplicationContext.currentActivity

  override fun getName(): String = "RNFB${moduleName}Module"

  override fun getConstants(): MutableMap<String, Any> = HashMap()

  companion object {
    @JvmStatic
    fun rejectPromiseWithExceptionMap(
      promise: Promise?,
      exception: Exception,
    ) {
      val exceptionMap = SharedUtils.getExceptionMap(exception)
      promise!!.reject(exception, exceptionMap)
    }

    @JvmStatic
    fun rejectPromiseWithCodeAndMessage(
      promise: Promise?,
      code: String?,
      message: String?,
      resolver: ReadableMap?,
    ) {
      val userInfoMap = Arguments.createMap()
      userInfoMap.putString("code", code)
      userInfoMap.putString("message", message)
      if (resolver != null) {
        userInfoMap.putMap("resolver", resolver)
      }
      promise!!.reject(code, message, userInfoMap)
    }

    @JvmStatic
    fun rejectPromiseWithCodeAndMessage(
      promise: Promise?,
      code: String?,
      message: String?,
    ) {
      val userInfoMap = Arguments.createMap()
      userInfoMap.putString("code", code)
      userInfoMap.putString("message", message)
      promise!!.reject(code, message, userInfoMap)
    }

    @JvmStatic
    fun rejectPromiseWithCodeAndMessage(
      promise: Promise?,
      code: String?,
      message: String?,
      nativeErrorMessage: String?,
    ) {
      val userInfoMap = Arguments.createMap()
      userInfoMap.putString("code", code)
      userInfoMap.putString("message", message)
      userInfoMap.putString("nativeErrorMessage", nativeErrorMessage)
      promise!!.reject(code, message, userInfoMap)
    }
  }
}
